service/worker/scheduler/activities.go

```diff
@@ -326,3 +295,97 @@ func translateError(err error, msgPrefix string) error {
+		workflowStatus:          workflowStatus,
+		logger:                  logger,
+		resultStorageNumberSize: resultStorageSize,
+	}
+}
+
+//nolint:revive
+func (r responseBuilder) Build(event *historypb.HistoryEvent) (*schedspb.WatchWorkflowResponse, error) {
+	switch r.workflowStatus {
+	case enumspb.WORKFLOW_EXECUTION_STATUS_RUNNING:
+		if r.request.LongPoll {
+			return nil, errTryAgain // not closed yet, just try again
+		}
+		return r.makeResponse(nil, nil), nil
+	case enumspb.WORKFLOW_EXECUTION_STATUS_COMPLETED:
+		if attrs := event.GetWorkflowExecutionCompletedEventAttributes(); attrs == nil {
+			return nil, errNoAttrs
+		} else if len(attrs.NewExecutionRunId) > 0 {
+			// this shouldn't happen because we don't allow old-cron workflows as scheduled, but follow it anyway
+			return nil, errFollow(attrs.NewExecutionRunId)
+		} else {
+			result := attrs.Result
+			if r.isTooBig(result) {
+				r.logger.Error(
+					fmt.Sprintf("result dropped due to its size %d", proto.Size(result)),
+					tag.WorkflowID(r.request.Execution.WorkflowId))
+				result = nil
+			}
+			return r.makeResponse(result, nil), nil
+		}
+	case enumspb.WORKFLOW_EXECUTION_STATUS_FAILED:
+		if attrs := event.GetWorkflowExecutionFailedEventAttributes(); attrs == nil {
+			return nil, errNoAttrs
+		} else if len(attrs.NewExecutionRunId) > 0 {
+			return nil, errFollow(attrs.NewExecutionRunId)
+		} else {
+			failure := attrs.Failure
+			if r.isTooBig(failure) {
+				r.logger.Error(
+					fmt.Sprintf("failure dropped due to its size %d", proto.Size(failure)),
+					tag.WorkflowID(r.request.Execution.WorkflowId))
+				failure = nil
+			}
+			return r.makeResponse(nil, failure), nil
+		}
+	case enumspb.WORKFLOW_EXECUTION_STATUS_CANCELED, enumspb.WORKFLOW_EXECUTION_STATUS_TERMINATED:
+		return r.makeResponse(nil, nil), nil
+	case enumspb.WORKFLOW_EXECUTION_STATUS_CONTINUED_AS_NEW:
+		if attrs := event.GetWorkflowExecutionContinuedAsNewEventAttributes(); attrs == nil {
+			return nil, errNoAttrs
+		} else {
+			return nil, errFollow(attrs.NewExecutionRunId)
+		}
+	case enumspb.WORKFLOW_EXECUTION_STATUS_TIMED_OUT:
+		if attrs := event.GetWorkflowExecutionTimedOutEventAttributes(); attrs == nil {
+			return nil, errNoAttrs
+		} else if len(attrs.NewExecutionRunId) > 0 {
+			return nil, errFollow(attrs.NewExecutionRunId)
+		} else {
+			return r.makeResponse(nil, nil), nil
+		}
+	}
+	return nil, errUnkownWorkflowStatus
+}
+
+func (r responseBuilder) isTooBig(m proto.Message) bool {
+	return proto.Size(m) > r.resultStorageNumberSize
+}
+
+func (r responseBuilder) makeResponse(result *commonpb.Payloads, failure *failurepb.Failure) *schedspb.WatchWorkflowResponse {
+	res := &schedspb.WatchWorkflowResponse{Status: r.workflowStatus}
+	if result != nil {
+		res.ResultFailure = &schedspb.WatchWorkflowResponse_Result{Result: result}
+	} else if failure != nil {
+		res.ResultFailure = &schedspb.WatchWorkflowResponse_Failure{Failure: failure}
+	}
+	return res
```

Comment:
```suggestion
	resp := &schedspb.WatchWorkflowResponse{Status: r.workflowStatus}
	if result != nil {
		resp.ResultFailure = &schedspb.WatchWorkflowResponse_Result{Result: result}
	} else if failure != nil {
		resp.ResultFailure = &schedspb.WatchWorkflowResponse_Failure{Failure: failure}
	}
	return resp
```

Calling it `resp` would make it clearer that it's different from the `result` argument.